### ISSUE TITLE
Unit Test Passing in repl, failing in terminal/CI

### ISSUE THREAD

**Issue by adam-james-v** [metabase/metabase#33394]
### Describe the bug

I've been working on `GET api/user` trying to change its behaviour to only show a list of users to admins and group managers (the list is limited to any groups they are managing).

As part of that change I want to modify some tests in `metabase.api.user-test` to reflect the different behaviour. As I worked on modifiying `user-list-test` I was able to have the test pass in the repl locally. I ran the test in `metabase.api.user-test` ns with: 

`(run-test user-list-test)`. 

Then, running in both CI and locally on my terminal I see test failures from that test. Poking around, I still don't know what's going on, but I have done the following to try narrow things down:

### To Reproduce

1. temporarily change the endpoint code in `metabase.api.user`:

```clojure
(api/defendpoint-schema GET "/"
  [status query group_id include_deactivated]
    {status              (s/maybe s/Str)
     query               (s/maybe s/Str)
     group_id            (s/maybe su/IntGreaterThanZero)
     include_deactivated (s/maybe su/BooleanString)}
  {:user_id          api/*current-user-id*
   :is_group_manager api/*is-group-manager?*})
```

The changes are:
 - remove any validation
 - just return a map of some values, nothing else.

Obviously, that's not the real endpoint, it's just for repro purposes.

2. Create a new test:

```clojure
(deftest asdf
  (testing "Things are going wrong in Terminal????."
    (t2.with-temp/with-temp
        [:model/PermissionsGroup           g {:name "Test Group"}
         :model/PermissionsGroupMembership a {:user_id (mt/user->id :rasta) :group_id (u/the-id g) :is_group_manager true}
         :model/PermissionsGroupMembership _ {:user_id (mt/user->id :lucky) :group_id (u/the-id g) :is_group_manager false}]
      (is (= (mt/user-http-request :rasta :get 200 "user")
             (select-keys a [:user_id :is_group_manager]))))))
```

So, the test will pass if the HTTP request returns the same `user_id` and `is_group_manager` boolean and fail otherwise.

It's probably wise to temporarily make a new user-test file and have only that test in there (with the necessary requires and fixtures), but I just commented all the deftests out as I was hacking on this.

3. Run the test in my repl.

I see this:

```shell
metabase.api.user-test> (run-test asdf)

Testing metabase.api.user-test
2023-08-22 00:05:13,264 DEBUG middleware.log :: GET /api/user 200 581.2 µs (0 DB calls) App DB connections: 0/4 Jetty threads: 5/50 (2 idle, 0 queued) (66 total active threads) Queries in flight: 0 (0 queued)

Ran 1 tests containing 2 assertions.
0 failures, 0 errors.
{:test 1, :pass 2, :fail 0, :error 0, :type :summary}
```

Things pass!

4. Run tests in the terminal.

`clj -X:dev:test :only metabase.api.user-test`

I see:

```shell
FAIL in metabase.api.user-test/asdf (user_test.clj:67)
Things are going wrong in Terminal????.
with temporary :model/PermissionsGroup with attributes
{:name "Test Group"}


with temporary :model/PermissionsGroupMembership with attributes
{:user_id 1, :group_id 3, :is_group_manager true}


with temporary :model/PermissionsGroupMembership with attributes
{:user_id 4, :group_id 3, :is_group_manager false}


expected: {:is_group_manager false, :user_id 1}
  actual: {:is_group_manager true, :user_id 1}
    diff: - {:is_group_manager false}
          + {:is_group_manager true}
```

For some reason, running in the terminal has different results. I do not understand why this happens yet.

My only theory so far is that this is somehow related to Group Management being an Enterprise feature? (is it, I can't recall?)

### Expected behavior

Tests should ideally pass in both the repl and in terminal.

### Information about your Metabase installation

```JSON
- running locally from master
```

### Severity

I can't seem to write a proper test so it's blocking another bugfix I was working on.

**Comment by adam-james-v** [comment]
@dpsutton I've tried to narrow things down to something passing in repl/failing in terminal as a starting point for this issue. This stems from our convo in Slack earlier.

**Comment by adam-james-v** [comment]
Ok, I've discovered the problem. It's not really a bug, I suppose, but I struggled with this for a while, so perhaps there's some way we could use tooling to hint away misunderstandings.

Anyway, here's what was wrong:

The test creates temporary group/groupMembership entities no problem, BUT since these features only have meaningful context with the enterprise feature `:advanced-permissions`, they won't do anything in a regular OSS setting. 

The test I was modifying actually was never correct, it was just passing circumstantially (because the user endpoint was always returning all users, even for non admins/group members.

So, the solution to this issue is:

```clojure
 (deftest asdf
  (testing "Things will now work in repl and terminal."
    (premium-features-test/with-premium-features #{:advanced-permissions}
     (t2.with-temp/with-temp
         [:model/PermissionsGroup           g {:name "Test Group"}
          :model/PermissionsGroupMembership a {:user_id (mt/user->id :rasta) :group_id (u/the-id g) :is_group_manager true}
          :model/PermissionsGroupMembership _ {:user_id (mt/user->id :lucky) :group_id (u/the-id g) :is_group_manager false}]
         (is (= (mt/user-http-request :rasta :get 200 "user")
                (select-keys a [:user_id :is_group_manager])))))))
```

This: `(premium-features-test/with-premium-features #{:advanced-permissions} ...)` was the crucial missing piece.

The test was passing in the repl because I am running the source with the enterprise features enabled, but the flag was not set in the test, so in the repl things run just fine, because the feature is already there, but that's just not the case when running this test in the terminal.
